enum class Some {
    FIRST,
    SECOND;
}

fun foo(s: Some) = <!NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER!>when (s) {
    FIRST -> <!UNRESOLVED_REFERENCE!>SECOND<!>
    SECOND -> <!UNRESOLVED_REFERENCE!>FIRST<!>
}<!>
