enum class Some {
    FIRST,
    SECOND;
}

fun foo(s: Some) = when (s) {
    FIRST -> <!ARGUMENT_TYPE_MISMATCH, UNRESOLVED_REFERENCE!>SECOND<!>
    SECOND -> <!ARGUMENT_TYPE_MISMATCH, UNRESOLVED_REFERENCE!>FIRST<!>
}
